org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/configurator/AbstractProjectConfigurator.java

```diff
@@ -154,28 +162,56 @@ public void setInitializationData(IConfigurationElement config, String propertyN
     this.name = config.getAttribute(ATTR_NAME);
   }
 
-  // TODO move to a helper
+  // TODO move to a helper, to which?
   public static void addNature(IProject project, String natureId, IProgressMonitor monitor) throws CoreException {
     addNature(project, natureId, IResource.KEEP_HISTORY, monitor);
   }
 
   /**
    * @since 1.3
    */
-  // TODO move to a helper
+  // TODO move to a helper, to which?
   public static void addNature(IProject project, String natureId, int updateFlags, IProgressMonitor monitor)
       throws CoreException {
     if(!project.hasNature(natureId)) {
       IProjectDescription description = project.getDescription();
-      String[] prevNatures = description.getNatureIds();
-      String[] newNatures = new String[prevNatures.length + 1];
-      System.arraycopy(prevNatures, 0, newNatures, 1, prevNatures.length);
-      newNatures[0] = natureId;
-      description.setNatureIds(newNatures);
+      var natures = Stream.concat(Stream.of(natureId), Arrays.stream(description.getNatureIds()));
+      description.setNatureIds(natures.toArray(String[]::new));
       project.setDescription(description, updateFlags, monitor);
     }
   }
 
+  /**
+   * Creates a problem marker in the pom.xml file at the specified element of the given Plugin-execution with the passed
+   * message and severity.
+   * <p>
+   * If the attribute of the execution is specified in a parent pom.xml outside of the workspace the marker is created
+   * at the parent-element of the request's project.
+   * </p>
+   * 
+   * @param execution the execution
+   * @param element the element to mark
+   * @param request the request of the project being build
+   * @param severity the problems severity, one of {@link IMarker#SEVERITY_INFO SEVERITY_INFO},
+   *          {@link IMarker#SEVERITY_WARNING SEVERITY_WARNING} or {@link IMarker#SEVERITY_ERROR SEVERITY_ERROR}
+   * @param message the message of the created problem marker
+   */
+  protected void createProblemMarker(MojoExecution execution, String element, ProjectConfigurationRequest request,
+      int severity, String message) {
+    SourceLocation location = SourceLocationHelper.findLocation(execution.getPlugin(), element);
+
+    String[] gav = location.getResourceId().split(":");
+    IMavenProjectFacade facade = projectManager.getMavenProject(gav[0], gav[1], gav[2]);
+    if(facade == null) {
+      // attribute specifying project (probably parent) is not in the workspace.
+      // The following code returns the location of the project's parent-element.
+      location = SourceLocationHelper.findLocation(request.mavenProject(), new MojoExecutionKey(execution));
+      facade = request.mavenProjectFacade();
+    }
+    MavenProblemInfo problem = new MavenProblemInfo(message, severity, location);
+    markerManager.addErrorMarker(facade.getPom(), IMavenConstants.MARKER_LIFECYCLEMAPPING_ID, problem);
```

Comment:
the marker should be made configurable (given as additional argument) as not every consumer uses that ID

Reply:
Also `IMavenConstants.MARKER_CONFIGURATION_ID` seems to be the more reasonable default for error with the configurator due to some mojo execution configuration.

Reply:
> Also `IMavenConstants.MARKER_CONFIGURATION_ID` seems to be the more reasonable default for error with the configurator due to some mojo execution configuration.

Good point.

> the marker id should be made configurable (given as additional argument) as not every consumer uses that ID

Do you want to use custom problem markers? 

One problem I see here, is that in most cases consumers are probably fine with the default problem marker. But `IMavenConstants.MARKER_CONFIGURATION_ID` is not part of the M2E's API so we should not encourage users to use it. Instead a overload of the method without marker-type that passes that type/id is probably more suitable. It should then make no claims about the type. If necessary we can just change the default in the future.

Reply:
If users want custom marker why can't they create them on their own?

Reply:
> If users want custom marker why can't they create them on their own?

Most of the methods/classes used in that method are internal. So to create a custom marker one a consumer would have to rely on M2E internals or would have to copy a lot of other code to rely only on Eclipse/Maven API.

Reply:
Well custom markers are... custom ... so what from m2e would be required? The only thing I can think of is determine the line/column in the XML and that could be public.

Reply:
> Well custom markers are... custom ... so what from m2e would be required? The only thing I can think of is determine the line/column in the XML and that could be public.

I think it is about the type/id String being custom. And from M2E the convenient way to add markers to a pom is wanted.
We could just make a method public that computes the location. But since the classes that carry that information are internal as well, we end up with a much larger API than just adding a overload with one additional parameter.
But of course the location computation maybe is interesting for other task as well?

Reply:
> I think it is about the type/id String being custom. And from M2E the convenient way to add markers to a pom is wanted.

But why? Custom marker only make sense if you add custom attributes and handle those custom, otherwise one can simply reuse the m2e one.

> we end up with a much larger API than just adding a overload with one additional parameter.

There are already many parameters and adding even more for unclear use-case do not make it better I think. All one needs would probably be in `SourceLocationHelper` already...

Reply:
OK, that's a point.
Maybe @kwin can elaborate a bit more about his use case and what would fit?

Nevertheless a `createProblemMarker()` method that creates `IMavenConstants.MARKER_CONFIGURATION_ID` problems still seems reasonable for me.

So if suitable we can still create an API around `SourceLocationHelper` and `SourceLocation` later.

Reply:
> But why? Custom marker only make sense if you add custom attributes and handle those custom, otherwise one can simply reuse the m2e one.

Custom markers appear in a dedicated section. There is not necessarily the need to set attributes on those. Also custom markers can be easily queried or cleared. So IMHO just exposing the marker ID as additional argument doesn't do any harm and allows this API to cover more use cases.

Reply:
OK sounds like there is a use-case for that.
Instead of having more and more arguments to that method, what do you think about encapsulating the problem data into a record with static factories?

The components would be `int problemSeverity, String problemMessage, String markerId` and two static factory methods
- `create(int problemSeverity, String problemMessage)`
- `create(int problemSeverity, String problemMessage, String markerId)`

The first one uses `IMavenConstants.MARKER_CONFIGURATION_ID` as `markerId`.
The java-doc of the record should then state that only canonical constructor is considered internal and should not be called by clients (maybe with `@noinstanciate`) and only the static factories should be used.

Reply:
Sounds good to me.